More GIFTI fixes

bcipolli opened the pull request:
This is an old branch that I never got to submitting. More commits to make `GIFTI` more pythonic and test the functionality a bit more thoroughly.

effigies commented on nibabel/gifti/tests/test_gifti.py:
```diff
@@ -11,8 +11,7 @@
 
-from numpy.testing import (assert_array_almost_equal,
-                           assert_array_equal)
+from numpy.testing import (assert_array_almost_equal, assert_array_equal)
```
Drop the parentheses.

The commit message makes bc017aa0713193969ecffcca2c7a82a421fbe1e8 look like it must have other stuff rebased out. Maybe re-write that one to match what's left?

effigies commented on nibabel/gifti/parse_gifti_fast.py:
```diff
@@ -168,13 +168,14 @@ def StartElementHandler(self, name, attrs):
             if "Dimensionality" in attrs:
-                self.da.num_dim = int(attrs["Dimensionality"])
-            for i in range(self.da.num_dim):
+                self.num_dim = int(attrs["Dimensionality"])
```
I think this is wrong. `self.num_dim` is now a property and can't be assigned.

bcipolli replied:
Yeah, confusing--it's set on the parser. The parser then asserts that the explicit value matches what was found in the data, while parsing. It's never actually set on the image.

effigies replied:
Oh, I see. Okay, so should you maybe set `self.num_dim = 0` at some point, so that `for i in range(self.num_dim):` doesn't fail if `"Dimensionality" not in attrs`?

Maybe replace the `if` block with:

    self.num_dim = int(attrs.get("Dimensionality", 0))

This is assuming `attrs` has a proper `dict` interface.

bcipolli replied:
Cool. Updated and pushed. Tests passed locally, but my battery will die before I can review things on my own. Feel free to wait until I'm able to review myself!

bcipolli commented:
OK, I've looked this over and made tweaks so that tests pass. I'm going to ignore the 0.004% decrease in test coverage (due to decreasing the total amount of code lol); I think this is ready to go!

effigies commented on nibabel/gifti/parse_gifti_fast.py:
```diff
@@ -167,14 +167,14 @@ def StartElementHandler(self, name, attrs):
             # dimensionality has to correspond to the number of DimX given
-            assert len(self.da.dims) == self.da.num_dim
+            # TODO (bcipolli): don't assert; raise parse warning, and recover.
+            assert len(self.da.dims) == self.num_dim
```
Sorry to prolong this. Unless I'm missing something, `self.num_dim` is never referenced outside this function. Any reason not to make it a local variable?

bcipolli replied:
No problem at all, thanks for your eyes on it! Good point; it was a member variable and I didn't think hard enough to realize it could be made local now. Will take care of that!

effigies commented on nibabel/gifti/gifti.py:
```diff
@@ -238,19 +241,37 @@ class GiftiDataArray(xml.XmlSerializable):
+        self.datatype = 0  # required attribute, void/none
+        self.ind_ord = 2  # python/numpy default: column major order
+        self.encoding = None
+        self.endian = None
```
Maybe these should be:

    self.encoding = encoding
    self.endian = endian

effigies commented on nibabel/gifti/gifti.py:
```diff
@@ -238,19 +241,37 @@ class GiftiDataArray(xml.XmlSerializable):
+                 endian=sys.byteorder,
+                 coordsys=None,
+                 ordering="C",
+                 meta=None):
```
None of these parameters are actually used except for data.

effigies commented on nibabel/gifti/parse_gifti_fast.py:
```diff
@@ -167,14 +167,14 @@ def StartElementHandler(self, name, attrs):
             # dimensionality has to correspond to the number of DimX given
-            assert len(self.da.dims) == self.da.num_dim
+            # TODO (bcipolli): don't assert; raise parse warning, and recover.
+            assert len(self.da.dims) == num_dim
```
Do you want to go ahead and raise the warning, or is that for a future iteration? (Given that functionality isn't changing, I'm okay with kicking this one down the road.)

effigies commented on nibabel/gifti/gifti.py:
```diff
@@ -68,10 +68,10 @@ def print_summary(self):
     name = str
     value = str
-
+    """
```
For these, if possible, I'd rather see proper docstrings than simply commenting these out.

At simplest, a quick description and a type description, e.g.:

    class GiftiNVPairs(object):
        """Name-value pair objects

        Instance variables
        ------------------
        name    : str
        value   : str
        """

Documenting instance variables doesn't seem to be standard nibabel practice, so I'm also open to just removing these. Still, seems a little bit of a shame to lose the type information altogether. Perhaps @matthew-brett has a feeling on best direction to go with these?

matthew-brett replied:
Yes, I agree, it's good to document these.   I just looked that up in https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#class-docstring

So I think this should be:

```
class GiftiNVPairs(object):
    """Name-value pair objects

    Attributes
    -------------
    name : str
        Maybe some text.
    value   : str
        Might be text here too.
    """
```

effigies commented on nibabel/gifti/gifti.py:
```diff
@@ -289,7 +311,6 @@ def from_array(klass,
-        cda.num_dim = len(darray.shape)
         cda.dims = list(darray.shape)
         if datatype is None:
             cda.datatype = data_type_codes.code[darray.dtype]
```
Given that you've reworked `__init__`, this function wouldn't take much work to clean up so that it ends with something like:

    return klass(darray, encoding, endian, coordsys, ordering, meta)

bcipolli commented:
@effigies I appreciate the feedback and your timely code reviews--very much so. At the same time, I feel like we're going backwards here. 

First, it feels very odd to fix something, only to get new comments to change something else. I have no idea when this PR will complete because each time something change, something new is requested. I'd really much rather get all the feedback at once, then get one request every time I change something.

Second, more and more I feel your comments are expanding the scope of the original PR. This PR is an improvement on what exists, not to scrub clean any code that I touch. The more the scope expands, the less interested I am to complete the PR or, to be frank, to contribute at all.

I agree that Gifti (and for that matter, many Nibabel objects) could use better documentation. I disagree that this needs to happen in this PR. I have much higher priorities with nibabel than writing docstrings for Gifti, and going through these intensive code reviews is actually preventing me from getting there.

So for me, either we need to make it easier to contribute code to nibabel, or I need to solve my problems without contributing back the code. But I can't spend this much of my time working and reworking and reworking existing code. I appreciate your and @matthew-brett 's willingness to use your time in that way, but I really don't have the time for it any longer, and need to get to the research quicker than what' been happening here.

Thanks!

matthew-brett commented:
HI Ben - it sounds like we could do a better job of helping you here.

Could you give us an idea of how we could do that?  I hear your feeling that progress is too slow, and I agree, for which I am clearly to blame.   I guess your frustration is partly the fact that I'm sometimes being responsive, and sometimes not, therefore slowing you down. So, I'll try very hard to keep right on top of stuff this week.  Expect me to respond within a couple of hours (I have a few meetings this week).  Also, please feel free to hit me on gchat if you want faster feedback.

But for specifics, I think it's completely reasonable for us not to ask you to clean code you haven't touched yourself, and please just say when we're doing that and you don't think it's reasonable.

Also please feel free to say when you'd rather not make a change, that's a reasonable answer, reviewing, as you know, is hard task, and we need feedback to do it properly.

Let's get this finished up in the next couple of days.

Lastly - Chris - I'd like to add my deep thanks for all your help - I really appreciate it.

effigies commented:
@bcipolli I appreciate where you're coming from. So you understand my position: I'm generally looking at these as I get a spare moment (or need to procrastinate on something) and making a comment as I see something. If you would rather that I take more time and present my thoughts all at once to avoid the back-and-forth thing, I can definitely do that.

Also, my assumption (now that the blitz of PRs from last year is over) is that @matthew-brett is the primary reviewer. This is not my project, and I feel uncomfortable okaying changes that might ultimately make things less clean. Hence the feeling-out of whether you feel like we're moving out-of-scope and deference to Matthew's opinions. I'm trying to ease the burden on both of you, because you've been doing a ton of work and he's keeping this whole thing afloat, but if I'm actually just annoying, I can step back.

So maybe in the short term (this PR), I'd recommend you take my comments as suggestions at best and sync on Matthew's time-scale. In the long-term, I hope that we can settle on a non-alienating review process.

matthew-brett commented:
On a little reflection - two thoughts:

* Maybe it would be useful to have a 20 minute hangout tomorrow sometime.  Are you both around?  I'm free anytime except 4 til 6 pacific time.
* If we're getting stuck, one option is to merge this and follow up with pull-requests to the merged code.

Chris - sorry you got stuck with me off doing other stuff and not paying proper attention - that's a tough position to be in, and thanks for being patient.

effigies commented:
Hangouts in general are hard for me, as I work in a pretty open lab where I'd be bothering people. If you set up a time that works for you two, I can see if I can make it work. Otherwise I'm happy to defer to your decisions.

No objections to merging this PR as is.

matthew-brett commented:
Ben - what do you think?

bcipolli commented:
Sorry--limited time these days, juggling a lot of things, hence being more strict about my time.

The hard thing for me is budgeting my time. @effigies perhaps if you're working through a PR piecemeal, just indicate so. I would simply lay off making any changes until you finish, so I can run through them more efficiently.

I'd also suggest that, if all of our time is so constrained, we might want to loosen up the code reviews. They're time-consuming for all involved, and I feel they're a major bottleneck. I'd rather write more automated tests than spend so much time iterating on style... but that's just my feeling.

@matthew-brett Sorry for not catching your suggestion for the hangout earlier. I'm at a conference in DC, and juggling a few things in addition, so ... hard for me to make these days. I'll be home Thursday/Friday if y'all want to chat.

My goal is, and always has been, to get CIFTI support in so that I can work with HCP data in Python. Hope we can get there soon!

matthew-brett commented:
Yes, I think we can get there soon.  And this is a good moment to take stock of the review process and try and improve it.

What would you like me to do?  I am free tomorrow, and I will be reviewing another huge nibabel pull request (#391), but I can also go over the cifti one.  That way you'd have all or nearly all pertinent comments for Thursday.

The alternatives are:

* merge as is and me / Chris submit PRs for you to review;
* wait until Thursday / Friday, chat about it on a hangout and proceed from there, planning to get the work merged after a couple of days.

bcipolli commented:
I think, just proceed as usual. Knowing that things are coming in pieces helps, and knowing that I can suggest to *not* implement some of the feedback helps too. I will just let things sit and wait until I hear the magic words--"Finished the review, and ... " :)

matthew-brett commented:
OK, fair enough.  I should have finished the review by end of the day tomorrow.

matthew-brett commented on nibabel/gifti/gifti.py:
```diff
@@ -104,6 +104,7 @@ def print_summary(self):
 
 
 class GiftiLabel(xml.XmlSerializable):
+    """
```
Ben - feel free to leave these, I'll submit another PR with docstring edits.

matthew-brett commented on nibabel/gifti/gifti.py:
```diff
@@ -238,19 +241,38 @@ class GiftiDataArray(xml.XmlSerializable):
+                 meta=None):
+        """
+        Returns a shell object that cannot be saved.
+        """
```
It would be good to have some docstring description for these guys.

bcipolli replied:
Cool. I have to say, I'm not clear enough on the `GIFTI` format to write these; I've simply been approaching `GIFTI` from the software engineering side.

Sounds like you're up for writing these, @matthew-brett ?

matthew-brett commented on nibabel/gifti/tests/test_gifti.py:
```diff
@@ -34,10 +34,11 @@ def test_gifti_image():
     gi = GiftiImage()
     assert_equal(gi.numDA, 0)
 
-    da = GiftiDataArray(data='data')
```
It is worth also testing case of initializing from a string?

bcipolli replied:
Yes. This could be tested separately, without saving, to validate that properties are being set.

matthew-brett commented on nibabel/gifti/tests/test_gifti.py:
```diff
@@ -97,13 +97,15 @@ def test_labeltable():
-        warnings.filterwarnings('once', category=DeprecationWarning)
+        warnings.filterwarnings('always', category=DeprecationWarning)
         assert_equal(len(GiftiDataArray().get_metadata()), 0)
+        assert_equal(len(w), 1)
```
Nice

matthew-brett commented:
All finished from my PoV.   Thanks for being patient.  I'll email to the neuroimaging mailing list about ways of my being more responsive.   Please hold me to strictly a maximum two day delay on PR reviews for your stuff, especially for small changes like these.

bcipolli commented:
@effigies @matthew-brett Thanks for the reviews. I added back a string test, left docstrings as-is, and rebased on the latest master. I think that should do it?

matthew-brett commented:
Fine with me.  Chris - any comments? Otherwise I plan to merge tomorrow.

effigies commented:
Nope, I'm good. 

On February 12, 2016 11:38:34 PM EST, Matthew Brett <notifications@github.com> wrote:
>Fine with me.  Chris - any comments? Otherwise I plan to merge
>tomorrow.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nipy/nibabel/pull/403#issuecomment-183588200

-- 
Chris Markiewicz

matthew-brett commented:
OK - thanks to you both - merging.
